doc/fvconfig.1
.ds PN fvconfig

.TH fvconfig 1 "May 2012" "FlowVisor" "FlowVisor Manual"

.SH NAME
fvconfig \- commandline tool to administer FlowVisor

.SH AUTHOR

Rob Sherwood <rob.sherwood@stanford.edu>
 
Ali Al-Shabibi <ali.al-shabibi@stanford.edu>

.SH SYNOPSIS
.B fvconfig match config.json \fIdpid\fR \fImatch\fR

.B fvconfig load config.json

.B fvconfig chpasswd config.json \fIslicename\fR

.B fvconfig convert config.json

.B fvconfig query config.json \fIdpid\fR \fI[slicename]\fR

.B fvconfig generateCert \fI[hostname]\fR

.B fvconfig generate newconfig.json \fI[hostname]\fR \fI[admin_passwd]\fR \fI[of_port]\fR \fI[api_port]\fR

.SH DESCRIPTION
The
.B fvconfig
program is a command line tool for manipulating the flowvisor's configuration
directly, as opposed to through a running flowvisor instance (SEE fvctl(1)).
As such, fvconfig relies on standard file permissions instead of flowvisor
user permissions.  Also, fvconfig should 
.B NOT
be used to modify the configuration of a running flowvisor as your changes will be overwritten.  To modify
the configuration with fvconfig, you must kill the flowvisor first, and the changes will take effect
once the flowvisor is restarted.

.SH COMMANDS

.TP
\fBmatch\fR
    Used for debugging, asks the flowvisor to list which rules match the given match 
    structure.  The format of the match is as described in fvctl(1). 

.TP
\fBload\fR
    Can be used to prepopulate the configuration from the given file. Note that
    the configuration in the database, if any, will be overwritten. 

.TP
\fBchpasswd\fR \fIslicename\fR
    Changes the password for a slice.  Useful for resetting the admin password
    if it is forgotten.  This command does modify the configuration, so should
        not be used on a running flowvisor.

.TP
\fBquery\fR \fIdpid\fR \fI[slicename]\fR
    Two useful debugging commands.  If the slice name is not specified, this command will
    list the slices that have the specified DPID in their slice.  If the slice name is 
    specified, it will list the ports of the DPID that are used by this slice.

.TP
\fBconvert\fR
    Converts old json configuration file to a json file. The output filename is the same but the extension is changed to .json.

.TP
\fBgenerateCert\fR \fI[hostname]\fR
    This command generates a new SSL Key store (mySSLKeyStore) with \fIhostname\fR as the
    SSL common name.  If no hostname is specified, the output from /bin/hostname is used.
.TP
\fBgenerate\fR  newconfig.json \fI[hostname]\fR \fI[admin_passwd]\fR \fI[of_port]\fR \fI[api_port]\fR
    Generate a new config.  This command also generates a new
    configuration file, complete with example slices Alice and Bob.
    FVconfig will prompt for a password for the initial admin account
    if one is not specified on the command-line.  The config listens on
    port \fIof_port\fR (default: 6633) for OpenFlow switches and port
    \fIapi_port\fR for the jsonRPC server (default: 8080).  Likely, if
    this config file is actually to be used, the first thing to do would
    be to delete the Alice and Bob slices, and then populate it with
    your own slices.  This command also generates a new SSL Key store
    (mySSLKeyStore) with hostname as the SSL common name (i.e., same as
    'generateCert').

.SH "SEE ALSO"

.BR flowvisor (8),
.BR fvctl (1)


RELEASE-NOTES
Flowvisor 0.8.5 : July 11 2012
    * FLOWVISOR-177: fvctl dumpConfig should output the config file in fvctl's working directory
    * FLOWVISOR-178: Slice creation renders flowvisor unresponsive in some cases
    * FLOWVISOR-179: Need callback when slice controller disconnects

Flowvisor 0.8.4 : June 22 2012
    * FLOWVISOR-175: Provide API to change slice password.
    * FLOWVISOR-173: FV rejects a Flow_mod request with DELETE_FLOW if the buffer_id is 0.
    * FLOWVISOR-167: Need an RPC API to allow db backup
    * FLOWVISOR-166: Update API to replace setConfig/getConfig
    * FLOWVISOR-165: FlowRule ids change when the flowspace is manipulated
    * FLOWVISOR-161: Checkpointing broken on FlowSpace modifications    
    * FLOWVISOR-160: Prevent NoMatch exception from bubbling up to the output
    * FLOWVISOR-159: Flowvisor crashes on empty flowspaces if there are controller connected

Flowvisor 0.8.3 : May 20 2012
    * Database-based flowspace storage
    * Config storage transitioned to JSON
    * Dump config file via API call
    * FLOWVISOR-104: writing config after each change to disk hangs flowvisors with large flowspace
    
Flowvisor 0.8.2 : Febuary 2 2012
    * Slice configuration parameter for controlling LLDP spam
    * JSON callbacks for fine-grained topology changes
    * FLOWVISOR-128: Warn users when adding flowspace rules for non-existing slices
    * FLOWVISOR-126: Slices containing '<' can not be deleted
    * FLOWVISOR-122: fvctl no longer supports registration of topology callbacks with authorization information
    * FLOWVISOR-121: XMLRPC web server logs to stderr instead of FVLog
    * FLOWVISOR-95: when slicing actions, the errors that bubble up are confusing
    * FLOWVISOR-73: Buffer ID sanitizing
    * FLOWVISOR-22: add config option for drop policy

Flowvisor 0.8.1 : July 22 2011
    * Better logging to find delays and process starvation
    * Changed default GC to be mark/sweep
    * Source documentation improvements
    * FLOWVISOR-120: disabled DNS lookup for stats_desc
    * FLOWVISOR-118: fixed FvSlicer memleak in XID translation
    * FLOWVISOR-117: bind fails for override IPv4 stacks
    * FLOWVISOR-116: move configfile read before logging

FlowVisor 0.8.0 : July 13 2011
    * fixed problem with FlowIDs exhausting
    * fixed memleaks that caused instability
    * fixed dynamic ports and topology discovery issues
    * fixed bugs in Vendor Stats and sendlen handling
    * updated developer documentation
    * added better support for remote debugging and JVM tweaks
    * fixed IPv6 dual-stack local port binding bug
    * changed superuser from 'root' to 'fvadmin'
    * lots of bugs fixed and sanity checks implemented

FlowVisor 0.7.2 : Mar 17 2011
    * added changeSlice fvctl/API command
    * fixed big bug in drop rules for disconnected controllers
    * fixed Null Pointer Exception race condition; stops crashing
    * added MRTG formatting scripts (thanks Chris Tengi!)

FlowVisor 0.7.1 : Feb 22 2011
    * diabled flow tracking by default
    * fixed bug in getSwitchFlowDB

FlowVisor 0.7 : 
    * fixed big I/O bug that caused switch flapping
    * added msg statistics, see `fvctl get{Slice,Switch}stats`
    * added flow tracking, see `fvctl get{Switch,SliceRewrite}FlowDB`
        to enable, run `fvctl setConfig flowvisor\!track_flows true`
    * lots of performance improvements
        * flow setup time ~5x improved
    * lots of small bugs fixed
    * password prompt no longer echos plaintext
    * better error handling in create slice and change flowspace

FlowVisor 0.6 - total rewrite (in Java!)
    * Single, unified flowspace config, not divided by slice
    * lldp is now rewritten with a trailer, not by rewritting the dst mac
    * OFError msgs now correctly encap the offending packet
    * NEW FlowVisor UserAPI: query and manipulate slice info


src/org/flowvisor/config/FVConfig.java
/**
 *
 */
package org.flowvisor.config;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.Console;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.LinkedList;

import org.flowvisor.api.APIAuth;
import org.flowvisor.exceptions.DuplicateControllerException;
import org.flowvisor.flows.FlowMap;

import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonToken;
import com.google.gson.stream.JsonWriter;

/**
 * Central collection of all configuration and policy information, e.g., slice
 * permissions, what port to run on, etc.
 *
 * Uses get/set on a hierarchy of nodes like sysctl,snmp, etc.
 * getInt("flowvisor.list_port") --> 6633
 * setString("slice.alice.controller_hostname"
 * ,"alice-controller.controllers.org")
 *
 * All of the set* operations will dynamically create the entry if it does not
 * exist.
 *
 * @author capveg
 *
 */
public class FVConfig {
	public static final String FS = "!";
	
	public static final String SLICE_SALT = "passwd_salt";
	public static final String SLICE_CRYPT = "passwd_crypt";

	public static final String SUPER_USER = "fvadmin";
	
	final static public int OFP_TCP_PORT = 6633;
	public static final long DelayWarning = 10;

	/**
	 * Return the flowmap associated with this node
	 *
	 * @param node
	 * @return
	 * @throws ConfigError
	 */
	static public FlowMap getFlowMap() throws ConfigError{
		FlowSpace proxy = FlowSpaceImpl.getProxy();
		return proxy.getFlowMap();
	}

	static synchronized public FlowMap getFlowSpaceFlowMap() {
		FlowSpace proxy = FlowSpaceImpl.getProxy();
		try {
			return proxy.getFlowMap();
		} catch (ConfigError e) {
			e.printStackTrace();
			throw new RuntimeException("WTF!?!  No FlowSpace defined!?!");
		}
		
	}

	/**
	 * Set the flowmap at this entry, creating it if it does not exist
	 *
	 * @param node
	 * @param val
	 * @throws ConfigError
	 */
	static public void setFlowMap(FlowMap val) throws ConfigError {
		FlowSpace proxy = FlowSpaceImpl.getProxy();
		proxy.setFlowMap(val);
	}


	/**
	 * Read XML-encoded config from filename
	 *
	 * @param filename
	 *            fully qualified or relative pathname
	 */
	public static synchronized void readFromFile(String filename)
			throws FileNotFoundException {
		
		JsonReader reader = new JsonReader(new BufferedReader(new InputStreamReader(new FileInputStream(filename))));
		reader.setLenient(true);
		try {
			
			String nextName = null;
			while (reader.hasNext()) {
				if (reader.peek() == JsonToken.END_DOCUMENT) 
					break;
				else if (reader.peek() == JsonToken.BEGIN_OBJECT)
					reader.beginObject();
				else if (reader.peek() == JsonToken.END_OBJECT) {
					
					reader.endObject();
				} else if (reader.peek() == JsonToken.NAME)
					 nextName = reader.nextName();
				
				if (nextName == null)
					continue;
				
				if (nextName.equalsIgnoreCase(Flowvisor.FLOWVISOR)) 
					FlowvisorImpl.getProxy().fromJson(reader);
				else if (nextName.equalsIgnoreCase(Slice.TSLICE)) 
					SliceImpl.getProxy().fromJson(reader);					
				else if (nextName.equalsIgnoreCase(FlowSpace.FS)) {
					reader.beginArray();
					FlowSpaceImpl.getProxy().fromJson(reader);
				} else if (nextName.equalsIgnoreCase(Switch.SWITCH)) {
					reader.beginArray();
					while (true) {
						reader.beginObject();
						String dpid = reader.nextName();
						if (dpid.equalsIgnoreCase("default")) {
							FlowvisorImpl.getProxy().setFloodPerm(reader.nextString());
						} else
							//FIXME currently does nothing!
							SwitchImpl.getProxy().fromJson(reader);
						reader.endObject();
						if (reader.peek() == JsonToken.END_ARRAY) 
							break;
					}
					reader.endArray();		
						
				} else
					throw new RuntimeException("What the hell is a " + nextName
							+ "?!? Stop screwing with the config file!");
			}
			
			
		} catch (IOException e) {
			System.err.println("Error while parsing config file " + e.getMessage());
		} finally {
			try {
				reader.close();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
		
		
	}

	
	/**
	 * Write XML-encoded config to filename
	 *
	 * @param filename
	 *            fully qualified or relative pathname
	 * @throws FileNotFoundException 
	 * @throws IOException 
	 */
	public static synchronized void writeToFile(String filename) throws FileNotFoundException  {
		
		JsonWriter writer = new JsonWriter(new BufferedWriter(new OutputStreamWriter(new FileOutputStream(filename))));
		writer.setIndent("   ");
		try {
			writer.beginObject();
			FlowvisorImpl.getProxy().toJson(writer);
			SliceImpl.getProxy().toJson(writer);
			FlowSpaceImpl.getProxy().toJson(writer);
			writer.endObject();
		} catch (IOException e) {
			System.err.println("Error whie writing config file " + e.getMessage());
		} finally {
			try {
				writer.close();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}
	
	public static void createSlice(String sliceName,
			String controller_hostname, int controller_port, String drop_policy, String passwd,
			String slice_email, String creatorSlice) throws InvalidSliceName,
			DuplicateControllerException {
		FVConfig.createSlice(sliceName, controller_hostname, controller_port, drop_policy,
				passwd, APIAuth.getSalt(), slice_email, creatorSlice);
	}

	public synchronized static void createSlice(String sliceName,
			String controller_hostname, int controller_port, String drop_policy, String passwd,
			String salt, String slice_email, String creatorSlice)
			throws InvalidSliceName, DuplicateControllerException {
		
		sliceName = FVConfig.sanitize(sliceName);
		Slice proxy = SliceImpl.getProxy();
		proxy.createSlice(sliceName, controller_hostname, controller_port,
				drop_policy, passwd, salt, slice_email, creatorSlice);
	}

	public static String readPasswd(String prompt) throws IOException {
		Console cons = System.console();
		if (cons != null) {
			char[] passwd = cons.readPassword(prompt);
			return new String(passwd);
		} else {
			/**
			 * This is a hack to get around the fact that in java,
			 * System.console() will return null if the calling process is not a
			 * tty, e.g., with `fvctl listSlices | less`
			 */
			System.err.print(prompt);
			BufferedReader reader = new BufferedReader(new InputStreamReader(
					System.in));
			return reader.readLine();
		}
	}
	
	// Slice config params

	public static synchronized void deleteSlice(String sliceName) throws InvalidSliceName {
		sliceName = FVConfig.sanitize(sliceName);
		Slice proxy = SliceImpl.getProxy();
		proxy.deleteSlice(sliceName);
	}
	
	public static void setSliceContactEmail(String sliceName, String email) throws ConfigError {
		Slice proxy = SliceImpl.getProxy();
		proxy.setContactEmail(sliceName, email);
		
	}
	
	public static void setSliceHost(String sliceName, String hostname) throws ConfigError {
		Slice proxy = SliceImpl.getProxy();
		proxy.setcontroller_hostname(sliceName, hostname);
	}
	
	public static void setSlicePort(String sliceName, Integer port) throws ConfigError{
		Slice proxy = SliceImpl.getProxy();
		proxy.setcontroller_port(sliceName, port);
	}
	
	public static void setSliceDropPolicy(String sliceName, String policy) {
		Slice proxy = SliceImpl.getProxy();
		proxy.setdrop_policy(sliceName, policy);
	}

	public static boolean checkSliceName(String sliceName) {
		Slice proxy = SliceImpl.getProxy();
		return proxy.checkSliceName(sliceName);
	}
	
	public static String getSliceEmail(String sliceName) throws ConfigError {
		Slice proxy = SliceImpl.getProxy();
		return proxy.getEmail(sliceName);
	}
	
	public static String getSlicePolicy(String sliceName) {
		Slice proxy = SliceImpl.getProxy();
		return proxy.getdrop_policy(sliceName);
	}
	
	public static Integer getSlicePort(String sliceName) throws ConfigError {
		Slice proxy = SliceImpl.getProxy();
		return proxy.getcontroller_port(sliceName);
	}
	
	public static String getSliceHost(String sliceName) throws ConfigError {
		Slice proxy = SliceImpl.getProxy();
		return proxy.getcontroller_hostname(sliceName);
	}
	
	public static String getPasswdElm(String sliceName, String elm) throws ConfigError {
		Slice proxy = SliceImpl.getProxy();
		return proxy.getPasswdElm(sliceName, elm);
	}
	
	public static String getSliceCreator(String sliceName) throws ConfigError {
		Slice proxy = SliceImpl.getProxy();
		return proxy.getCreator(sliceName);
		
	}
	
	public static LinkedList<String> getAllSlices() throws ConfigError {
		Slice proxy = SliceImpl.getProxy();
		return proxy.getAllSliceNames();
	}
	
	public static String getDropPolicy(String sliceName) throws ConfigError {
		Slice proxy = SliceImpl.getProxy();
		return proxy.getdrop_policy(sliceName);
	}
	
	public static Boolean getLLDPSpam(String sliceName) throws ConfigError {
		Slice proxy = SliceImpl.getProxy();
		return proxy.getlldp_spam(sliceName);
	}
	
	
	
	// Flowisor config params.
	public static int getAPIWSPort(int id) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getAPIWSPort(id);
	}
	
	public static int getAPIWSPort() throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getAPIWSPort();
	}
	
	public static int getJettyPort(int id) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getJettyPort(id);
	}
	
	public static int getJettyPort() throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getJettyPort();
	}
	
	public static int getListenPort(int id) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getListenPort(id);
	}

	
	public static int getListenPort() throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getListenPort();
	}
	
	public static boolean getFlowTracking(int id) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.gettrack_flows(id);
	}
	
	public static boolean getFlowTracking() throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.gettrack_flows();
	}
	
	public static boolean getCheckPoint(int id) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getCheckPoint(id);
	}
	
	public static boolean getCheckPoint() throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getCheckPoint();
	}
	
	public static boolean getStatsDescHack(int id) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getstats_desc_hack(id);
	}
	
	public static boolean getStatsDescHack() throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getstats_desc_hack();
	}
	
	public static String getLogging(int id) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getLogging(id);
	}
	
	public static String getLogging() throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getLogging();
	}
	
	public static String getLogFacility(int id) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getLogFacility(id);
	}
	
	public static String getLogFacility() throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getLogFacility();
	}
	
	public static String getLogIdent(int id) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getLogIdent(id);
	}
	
	public static String getLogIdent() throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getLogIdent();
	}
	
	public static String getFloodPerm(Long dpid) throws ConfigError {
		if (dpid == null)
			return FlowvisorImpl.getProxy().getFloodPerm();
		else
			return SwitchImpl.getProxy().getFloodPerm(dpid);
	}
	
	public static Boolean getTopologyServer(int id) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getTopologyServer(id);
	}
	
	public static Boolean getTopologyServer() throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		return proxy.getTopologyServer();
	}
	
	public static void setTopologyServer(int id, Boolean topo) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setTopologyServer(id, topo);
	}
	
	public static void setTopologyServer(Boolean topo) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setTopologyServer(topo);
	}
	
	public static void setLogging(int id, String logging) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setLogging(id, logging);
	}
	
	public static void setLogging(String logging) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setLogging(logging);
	}
	
	public static void setLogFacility(int id, String facility) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setLogFacility(id, facility);
	}
	
	public static void setLogFacility(String facility) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setLogFacility(facility);
	}
	
	public static void setLogIdent(int id, String ident) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setLogFacility(id, ident);
	}
	
	public static void setLogIdent(String ident) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setLogFacility(ident);
	}
	
	public static void setStatsDescHack(int id, boolean statsHack) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setstats_desc_hack(id, statsHack);
	}
	
	public static void setStatsDescHack(boolean statsHack) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setstats_desc_hack(statsHack);
	}
	
	public static void setFlowTracking(int id, boolean track) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.settrack_flows(id, track);
	}
	
	public static void setFlowTracking(boolean track) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.settrack_flows(track);
	}
	
	public static void setListenPort(Integer port) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setListenPort(port);
	}
	
	public static void setListenPort(int id, Integer port) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setListenPort(id, port);
	}
	
	public static void setAPIWSPort(Integer port) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setAPIWSPort(port);
	}
	
	public static void setAPIWSPort(int id, Integer port) throws ConfigError {
		Flowvisor proxy = FlowvisorImpl.getProxy();
		proxy.setAPIWSPort(id, port);
	}
	
	
	
	
	/*public static boolean confirm(String base) {
		return (lookup(base) != null);
	}*/

	/**
	 * Return the name of the super user account
	 *
	 * @return
	 */
	public static boolean isSupervisor(String user) {
		return SUPER_USER.equals(user);
	}

	/**
	 * Replace all non-kosher characters with underscores
	 *
	 * @param str
	 * @return
	 */
	public static String sanitize(String str) {
		return str.replaceAll("[^a-zA-Z0-9,_+=:-]", "-");
	}

	/**
	 * Create a default config db and write it to arg1
	 *
	 * @param args
	 *            filename
	 * @throws FileNotFoundException
	 * @throws ConfigError
	 * @throws NumberFormatException
	 */

	public static void main(String args[]) throws FileNotFoundException,
			IOException, NumberFormatException, ConfigError {
		if (args.length < 1) {
			System.err
					.println("Usage: FVConfig config.xml [fvadmin_passwd] [of_listen_port] [rpc_listen_port]");
			System.exit(1);
		}
		String filename = args[0];
		String passwd;
		
		if (args.length > 1)
			passwd = args[1];
		else
			passwd = FVConfig
					.readPasswd("Enter password for account 'fvadmin' on the flowvisor:");
		
		
		System.err.println("Generating default config in db");
		
		LoadConfig.defaultConfig(passwd);
		
		FVConfigurationController.init(new ConfDBHandler());
		// set the listen port, if requested
		if (args.length > 2)
			FVConfig.setListenPort(Integer.valueOf(args[2]));
			
		if (args.length > 3)
			FVConfig.setAPIWSPort(Integer.valueOf(args[3]));
			
		FVConfigurationController.init(new ConfDBHandler());
		
		
		System.err.println("Outputing config file " + filename);
		FVConfig.writeToFile(filename);
		
		
	}

	public synchronized static void setPasswd(String sliceName, String salt,
			String crypt) {
		
		Slice proxy = SliceImpl.getProxy();
		try {
			proxy.setPasswd(sliceName, salt, crypt);
		} catch (ConfigError e) {
			e.printStackTrace();
			throw new RuntimeException(e);
		}
		
	}

	
	
}
